pkg/bgp/speaker/speaker.go

```diff
@@ -247,43 +247,53 @@ func (s *MetalLBSpeaker) notifyNodeEvent(op Op, nodeMeta metaGetter, podCIDRs *[
+	go eventLoop[*slim_corev1.Node](ctx, s, lnr, nodePodCIDRs)
 }
 
-func (s *MetalLBSpeaker) OnUpdateNode(oldNode, newNode *slim_corev1.Node, swg *lock.StoppableWaitGroup) error {
-	return s.notifyNodeEvent(Update, newNode, nodePodCIDRs(newNode), false)
+func (s *MetalLBSpeaker) SubscribeToLocalCiliumNodeResource(ctx context.Context, lcnr agentK8s.LocalCiliumNodeResource) {
+	go eventLoop[*ciliumv2.CiliumNode](ctx, s, lcnr, ciliumNodePodCIDRs)
 }
 
-// OnDeleteNode notifies the Speaker of a node deletion.
-//
-// When the speaker discovers the node that it is running on
-// is shuttig down it will send a BGP message to its peer
-// instructing it to withdrawal all previously advertised
-// routes.
-func (s *MetalLBSpeaker) OnDeleteNode(node *slim_corev1.Node, swg *lock.StoppableWaitGroup) error {
-	return s.notifyNodeEvent(Delete, node, nodePodCIDRs(node), true)
+type metaGetterObject interface {
+	k8sRuntime.Object
+	metaGetter
 }
 
-// OnAddCiliumNode notifies the Speaker of a new CiliumNode.
-func (s *MetalLBSpeaker) OnAddCiliumNode(node *ciliumv2.CiliumNode, swg *lock.StoppableWaitGroup) error {
-	return s.notifyNodeEvent(Add, node, ciliumNodePodCIDRs(node), false)
-}
-
-// OnUpdateCiliumNode notifies the Speaker of an update to a CiliumNode.
-func (s *MetalLBSpeaker) OnUpdateCiliumNode(oldNode, newNode *ciliumv2.CiliumNode, swg *lock.StoppableWaitGroup) error {
-	return s.notifyNodeEvent(Update, newNode, ciliumNodePodCIDRs(newNode), false)
-}
-
-// OnDeleteCiliumNode notifies the Speaker of a CiliumNode deletion.
-//
-// When the speaker discovers the node that it is running on
-// is shuttig down it will send a BGP message to its peer
-// instructing it to withdrawal all previously advertised
-// routes.
-func (s *MetalLBSpeaker) OnDeleteCiliumNode(node *ciliumv2.CiliumNode, swg *lock.StoppableWaitGroup) error {
-	return s.notifyNodeEvent(Delete, node, ciliumNodePodCIDRs(node), true)
+func eventLoop[T metaGetterObject](ctx context.Context, s *MetalLBSpeaker, res resource.Resource[T], nodePodCIDRs func(T) *[]string) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Don't retry on errors, the downstream code isn't expecting it.
+	evs := res.Events(ctx, resource.WithErrorHandler(resource.RetryUpTo(0)))
+
+	var existing *T
+	for ev := range evs {
+		var err error
+		switch ev.Kind {
+		case resource.Upsert:
+			node := ev.Object
+			if existing == nil {
+				err = s.notifyNodeEvent(Add, node, nodePodCIDRs(node), false)
+			} else {
+				err = s.notifyNodeEvent(Update, node, nodePodCIDRs(node), false)
+			}
+			existing = &node
+		case resource.Delete:
+			// When the speaker discovers the node that it is running on
+			// is shuttig down it will send a BGP message to its peer
+			// instructing it to withdrawal all previously advertised
+			// routes.
+			node := ev.Object
+			err = s.notifyNodeEvent(Delete, node, nodePodCIDRs(node), true)
+		}
+		ev.Done(err)
```

Comment:
Resource[T] doesn't log failed events (as it doesn't have enough context, so it's left to subscribers). Could you check whether how this now differs from K8sWatcher and whether logging behavior changed?

Reply:
The logging behaviour did not change AFAICT - cilium node chain did not log by itself; and the metric of k8s event success is baked into resource, so no change there.